File: npc_sim/events/stimulus.py
# Copyright 2025-2026 Sadık Abdusselam Albayrak
# Licensed under the Apache License, Version 2.0
"""Raw world signal emitted by entities."""

from __future__ import annotations
from enum import Enum
from npc_sim.core.sim_vector3 import SimVector3


class StimulusType(Enum):
    VISUAL = "Visual"
    AUDIO = "Audio"
    SOCIAL = "Social"
    OLFACTORY = "Olfactory"


class Stimulus:
    """
    A raw world signal emitted by any entity (NPC, object, event).
    The StimulusDispatcher routes stimuli to nearby NPCs;
    each NPC's PerceptionSystem then filters them.
    """

    def __init__(
        self,
        stimulus_type: StimulusType,
        source_id: str,
        source_position: SimVector3,
        intensity: float,
        timestamp: float,
        tag: str = "",
        payload: object = None,
    ):
        self.type = stimulus_type
        self.source_id = source_id or ""
        self.source_position = source_position
        self.intensity = max(0.0, min(intensity, 1.0))
        self.timestamp = timestamp
        self.tag = tag or ""
        self.payload = payload

    def __repr__(self) -> str:
        return (f"[Stimulus] {self.type.value} '{self.tag}' from {self.source_id} "
                f"@ {self.source_position} I={self.intensity:.2f} t={self.timestamp:.1f}")

File: Runtime/NPC/Stimulus.cs
using System;

namespace ForgeProject.Sim
{
    /// <summary>Types of sensory stimuli emitted into the world.</summary>
    public enum StimulusType
    {
        Visual,
        Audio,
        Social,
        Olfactory
    }

    /// <summary>
    /// A raw world signal emitted by any entity (NPC, object, event).
    /// The <see cref="StimulusDispatcher"/> routes stimuli to nearby NPCs;
    /// each NPC's <see cref="PerceptionSystem"/> then filters them.
    /// </summary>
    [Serializable]
    public sealed class Stimulus
    {
        // ─── Identity ─────────────────────────────────────────────────────────────
        public string       SourceId       { get; }
        public StimulusType Type           { get; }

        // ─── Spatial ──────────────────────────────────────────────────────────────
        public SimVector3   SourcePosition { get; }

        // ─── Signal ───────────────────────────────────────────────────────────────
        /// <summary>Base intensity [0, 1]. Attenuated with distance by the dispatcher.</summary>
        public float        Intensity      { get; }

        /// <summary>
        /// Semantic tag that percept filters match against.
        /// Examples: "Threat", "Food", "Ally", "Noise", "Fire".
        /// </summary>
        public string       Tag            { get; }

        /// <summary>Simulation timestamp when this stimulus was emitted.</summary>
        public float        Timestamp      { get; }

        /// <summary>Optional arbitrary payload (ItemRef, NpcRef, SimEvent…).</summary>
        public object       Payload        { get; }

        // ─── Constructor ──────────────────────────────────────────────────────────

        public Stimulus(
            StimulusType type,
            string       sourceId,
            SimVector3   sourcePosition,
            float        intensity,
            float        timestamp,
            string       tag     = "",
            object       payload = null)
        {
            Type           = type;
            SourceId       = sourceId       ?? string.Empty;
            SourcePosition = sourcePosition;
            Intensity      = Math.Clamp(intensity, 0f, 1f);
            Timestamp      = timestamp;
            Tag            = tag            ?? string.Empty;
            Payload        = payload;
        }

        public override string ToString() =>
            $"[Stimulus] {Type} '{Tag}' from {SourceId} @ {SourcePosition} I={Intensity:F2} t={Timestamp:F1}";
    }
}
